src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala

```diff
@@ -209,26 +257,30 @@ object ArrayFunctions extends RegistryFunctions {
 
     registerIR("[:*]", TArray(tv("T")), TInt32()) { (a, i) =>
       val idx = genUID()
-      ArrayMap(
-        ArrayRange(I32(0),
-          If(ApplyComparisonOp(LT(TInt32()), i, I32(0)),
-            ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
-            i),
-          I32(1)),
-        idx,
-        ArrayRef(a, Ref(idx, TInt32())))
+      If(IsNA(a), a,
+        ArrayMap(
+          ArrayRange(
+            I32(0),
+            If(ApplyComparisonOp(LT(TInt32()), i, I32(0)),
+              ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
+              UtilFunctions.min(i, ArrayLen(a))),
+            I32(1)),
+          idx,
+          ArrayRef(a, Ref(idx, TInt32()))))
```

Comment:
I fixed all the array indexing functions to behave the same as python. See the tests for the intended behavior.